src/worker/rpc.rs

```diff
@@ -210,30 +210,29 @@ pub async fn run_worker(
 const MAX_RUNNING_DOWNLOADS: usize = 32;
 const MAX_ATTEMPTS: u32 = 8;
 
-async fn update_hw_state(state_ref: WorkerStateRef, hw_state_poll_interval: Duration) {
-    let mut poll_interval = tokio::time::interval(hw_state_poll_interval);
+async fn update_hw_state(
+    state_ref: WorkerStateRef,
+    hw_poll_interval: Duration,
+    mut hw_sampler: Result<HwSampler, psutil::Error>,
+) {
+    let mut poll_interval = tokio::time::interval(hw_poll_interval);
     loop {
-        let cpu_state = sample_cpu();
         poll_interval.tick().await;
-
-        match cpu_state {
-            Ok(state) => {
-                let cpu_usage_vec = state.done().unwrap_or_else(|error| {
-                    log::warn!("Error reading cpu_state!: {:?}", error);
-                    Vec::new()
-                });
-                let cpu_stats_vec: Vec<CPUStats> = cpu_usage_vec
-                    .into_iter()
-                    .map(|cpu_load| (cpu_load.into()))
-                    .collect();
-                state_ref
-                    .get_mut()
-                    .hardware_state
-                    .refresh_hw_state(cpu_stats_vec);
-                log::debug!("Hardware State Updated!");
+        match &mut hw_sampler {
+            Ok(sampler) => {
+                let hw_state = sampler.fetch_hw_state();
+                match hw_state {
+                    Ok(new_state) => {
+                        println!("Hware state = {:?}", new_state);
```

Comment:
```
Hware state = WorkerHwState { worker_cpu_usage: CPUStats { cpu_per_core_percent_usage: [0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0] }, worker_memory_usage: MemoryStats { total: 16446119936, free: 4725846016 }, worker_network_usage: NetworkStats { rx_bytes: 57536855, tx_bytes: 7797000, rx_packets: 74199, tx_packets: 43037, rx_errors: 0, tx_errors: 0 } }
Hware state = WorkerHwState { worker_cpu_usage: CPUStats { cpu_per_core_percent_usage: [4.9504952, 6.185567, 2.0, 3.030303, 7.8431373, 8.163265, 2.0408163, 4.0] }, worker_memory_usage: MemoryStats { total: 16446119936, free: 4745322496 }, worker_network_usage: NetworkStats { rx_bytes: 57536967, tx_bytes: 7797112, rx_packets: 74201, tx_packets: 43039, rx_errors: 0, tx_errors: 0 } }
```
The first cpu usage measurement is all 0's but I doubt anybody would send an overview request within 1 second of worker start

Reply:
I've removed this println in the next commit

Reply:
> The first cpu usage measurement is all 0's but I doubt anybody would send an overview request within 1 second of worker start

That's fine. But maybe we will need to store some bounded history of HW states, if overview lags for a while, so that the data is not lost.

Reply:
Aren't we planning to store the history on hq side?

Reply:
Long term history, yes. But for short termy Tako should probably also store a few records. The overview request can be delayed.

Reply:
So I'll store it as a vector of past states in the worker/state, with each overview request, this vector will be cleared. But I'd say let's do it in a subsequent pr and finish the dashboard first, i can do it right after we have a basic dashboard accepted

Reply:
Sounds good :+1: It should be a circular buffer instead of a vector, to have a bounded length and fast insertion on both ends. So e.g. `VecDeque`.